consensus/istanbul/uptime/uptime.go

```diff
@@ -0,0 +1,242 @@
+package uptime
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// Store provides a persistent storage for uptime entries
+type Store interface {
+	ReadAccumulatedEpochUptime(epoch uint64) *Uptime
+	WriteAccumulatedEpochUptime(epoch uint64, uptime *Uptime)
+}
+
+// Uptime contains the latest block for which uptime metrics were accounted for. It also contains
+// an array of Entries where the `i`th entry represents the uptime statistics of the `i`th validator
+// in the validator set for that epoch
+type Uptime struct {
+	LatestBlock uint64
+	Entries     []UptimeEntry
+}
+
+// UptimeEntry contains the uptime score of a validator during an epoch as well as the
+// last block they signed on
+type UptimeEntry struct {
+	ScoreTally      uint64
+	LastSignedBlock uint64
+}
+
+func (u *UptimeEntry) String() string {
+	return fmt.Sprintf("UptimeEntry { scoreTally: %v, lastBlock: %v}", u.ScoreTally, u.LastSignedBlock)
+}
+
+// GetUptimeMonitoringWindow retrieves the range [first block, last block] where uptime is to be monitored
+// for a give epoch. The range is inclusive.
+// First blocks of an epoch need to be skipped since we can't assess the last `lookbackWindow` block for validators
+// as those are froma different epoch.
+// Similarly, last block of epoch is skipped since we can't obtaine the signer for it; as they are in the next block
+func GetUptimeMonitoringWindow(epochNumber uint64, epochSize uint64, lookbackWindowSize uint64) (uint64, uint64) {
+	if epochNumber == 0 {
+		panic("no monitoring window for epoch 0")
+	}
+
+	epochFirstBlock, _ := istanbul.GetEpochFirstBlockNumber(epochNumber, epochSize)
+	epochLastBlock := istanbul.GetEpochLastBlockNumber(epochNumber, epochSize)
+
+	// first block to monitor:
+	// We need to wait for the completion of the first window with the start window's block being the
+	// 2nd block of the epoch, before we start tallying the validator score for epoch "epochNumber".
+	// We can't include the epoch's first block since it's aggregated parent seals
+	// is for the previous epoch's valset.
+	firstBlockToMonitor := epochFirstBlock + 1 + (lookbackWindowSize - 1)
```

Comment:
it's now `firstBlockToMonitor := newWindowStartingOn(epochFirstBlock, lookbackWindowSize).End` since the first valid block is the last block of the window starting at the beginning of the epoch